chalk-integration/src/lowering.rs

```diff
@@ -357,6 +376,52 @@ impl LowerProgram for Program {
                 Item::Clause(ref clause) => {
                     custom_clauses.extend(clause.lower_clause(&empty_env)?);
                 }
+                Item::OpaqueTyDefn(ref opaque_ty) => {
+                    if let Some(&opaque_ty_id) = opaque_ty_ids.get(&opaque_ty.identifier.str) {
+                        let parameter_kinds = opaque_ty
+                            .parameter_kinds
+                            .iter()
+                            .map(|k| k.lower())
+                            .collect::<Vec<_>>();
+
+                        let binders = empty_env.in_binders(parameter_kinds, |env| {
+                            let hidden_ty = opaque_ty.ty.lower(&env)?;
+
+                            let bounds: chalk_ir::Binders<Vec<chalk_ir::Binders<_>>> = env
```

Comment:
// Introduce a variable to represent the hidden "self type". This will be used in the bounds. So the `impl Trait<T1..Tn>` will be lowered to `exists<Self> { Self: Trait<T1..Tn> }`.